conda/common/url.py

```diff
@@ -125,11 +131,116 @@ def path_to_url(path):
     return path
 
 
+url_attrs = (
+    "scheme",
+    "path",
+    "query",
+    "fragment",
+    "username",
+    "password",
+    "hostname",
+    "port",
+)
+
+
+class Url(namedtuple("Url", url_attrs)):
+    """
+    Object used to represent a Url. The string representation of this object is a url string.
+
+    This object was inspired by the urllib3 implementation as it gives you a way to construct
+    URLs from various parts. The motivation behind this object was making something that is
+    interoperable with built the `urllib.parse.urlparse` function and has more features than
+    the built-in `ParseResult` object.
+    """
+
+    def __new__(
+        cls,
+        scheme=None,
+        path=None,
+        query=None,
+        fragment=None,
+        username=None,
+        password=None,
+        hostname=None,
+        port=None,
+    ):
+        if path and not path.startswith("/"):
+            path = "/" + path
+        if scheme:
+            scheme = scheme.lower()
+        if hostname:
+            hostname = hostname.lower()
+        return super(Url, cls).__new__(
+            cls, scheme, path, query, fragment, username, password, hostname, port
+        )
+
+    @property
+    def auth(self):
+        if self.username and self.password:
+            return f"{self.username}:{self.password}"
+        elif self.username:
+            return self.username
+
+    @property
+    def netloc(self):
+        if self.port:
+            return f"{self.hostname}:{self.port}"
+        return self.hostname
+
+    def __str__(self):
+        scheme, path, query, fragment, username, password, hostname, port = self
+        url = ""
+
+        if scheme:
+            url += f"{scheme}://"
+        if password and username:
+            url += f"{username}:{password}" + "@"
+        if hostname:
+            url += hostname
+        if port:
+            url += f":{port}"
+        if path:
+            url += path
+        if query:
+            url += f"?{query}"
+        if fragment:
+            url += f"#{fragment}"
+
+        return url
+
+    def as_dict(self) -> dict:
+        return {fld: getattr(self, fld) for fld in url_attrs}
+
+    def replace(self, **kwargs) -> "Url":
+        """
+        Allows for generating a new copy of a Url object.
+        This makes stripping username/password easier.
+
+        >>> u = Url(scheme='https', hostname='anaconda.com', path='/hello',
+        ...         username='user', password='pass')
+        >>> u1 = u.replace(username='', password='')
+        >>> str(u1)
+        'https://anaconda.com/hello'
+        """
+        new_kwargs = self.as_dict()
+        new_kwargs.update(kwargs)
+
+        return self.__class__(**new_kwargs)
```

Comment:
Hm, actually both `as_dict` and `replace` are predefined as part of [`collections.namedtuple`/`typing.NamedTuple`](https://docs.python.org/3/library/collections.html#collections.namedtuple). Should we just use those methods (`_asdict` and `_replace`) and/or alias them?

```suggestion
    replace = _Url._replace
```